### PR TITLE
Added subjectAltName to auto-generated self-signed hub certificate in postinstall scriptlet

### PR DESCRIPTION
**Pull request by craigcomstock**
Ticket: ENT-7598
Changelog: none

### CODE REVIEW

**Comment by craigcomstock** [comment]
getting a package here for ubu24 [Build Status](https://ci.cfengine.com/job/pr-pipeline/10929/)

File: packaging/common/cfengine-hub/postinstall.sh [hunk 1]
```diff
@@ -334,7 +334,7 @@ if [ ! -f $CFENGINE_MP_CERT ]; then
   ${CFENGINE_OPENSSL} req -utf8 -sha256 -nodes -new -subj "/CN=$CFENGINE_LOCALHOST" -key ${CFENGINE_MP_KEY} -out ${CFENGINE_MP_CSR} ${OPENSSL_CNF}
 
   # Generate CRT
-  ${CFENGINE_OPENSSL} x509 -req -days ${CFENGINE_SSL_DAYS_VALID} -in ${CFENGINE_MP_CSR} -signkey ${CFENGINE_MP_KEY} -out ${CFENGINE_MP_CERT}
+  ${CFENGINE_OPENSSL} x509 -req -days ${CFENGINE_SSL_DAYS_VALID} -in ${CFENGINE_MP_CSR} -signkey ${CFENGINE_MP_KEY} -out ${CFENGINE_MP_CERT} -extensions SAN -extfile <(cat $OPENSSL_CNF <(printf "[SAN]\nsubjectAltName=DNS:$CFENGINE_LOCALHOST"))
```
Comment (craigcomstock) [inline packaging/common/cfengine-hub/postinstall.sh:337]: probably be a good idea to add reasonable defaults as well like: 127.0.0.1, localhost, and $(hostname -i) output. Will do.

**Comment by craigcomstock** [comment]
try again, no subshells inside of printf or something like that :)

[Build Status](https://ci.cfengine.com/job/pr-pipeline/10930/)

**Comment by craigcomstock** [comment]
let's break it down to as simple as possible, POSIX style right?

[Build Status](https://ci.cfengine.com/job/pr-pipeline/10931/)

**Comment by craigcomstock** [comment]
one more time, for wacky centos-7 hostname -I/hostname -i output

[Build Status](https://ci.cfengine.com/job/pr-pipeline/10932/)

File: packaging/common/cfengine-hub/postinstall.sh [hunk 2]
```diff
@@ -333,8 +336,14 @@ if [ ! -f $CFENGINE_MP_CERT ]; then
   # Generate a CSR in ${CFENGINE_MP_CSR} with key ${CFENGINE_MP_KEY}
   ${CFENGINE_OPENSSL} req -utf8 -sha256 -nodes -new -subj "/CN=$CFENGINE_LOCALHOST" -key ${CFENGINE_MP_KEY} -out ${CFENGINE_MP_CSR} ${OPENSSL_CNF}
 
+  # Build configuration with reasonable default subjectAltName entries
+  rm -f "$CFENGINE_MP_SSLCONF"
+  test -f "$OPENSSL_CNF" && cat "$OPENSSL_CNF" > $CFENGINE_MP_SSLCONF
+  CFENGINE_LOCALIP=$(for i in $(hostname -i; hostname -I); do echo $i; done | sed '/::1/d' | sed '/127.0.0.1/d' | sort -u | xargs -n1 printf "IP:%s\n" | paste -sd "," -)
+  printf "[SAN]\nsubjectAltName=DNS:$CFENGINE_LOCALHOST,DNS:localhost,IP:127.0.0.1,$CFENGINE_LOCALIP" >> $CFENGINE_MP_SSLCONF
```
Comment (nickanderson) [inline packaging/common/cfengine-hub/postinstall.sh:343]: Why strip lines with 127.0.0.1 and then manually include them?

**Review by nickanderson — APPROVED** [review]
Seems ok to me. Don't quite grok why the long one-liner exactly, I see your just trying to get the local ips a host might have, but i see that it does work (checked it on a host with many various IPs).

File: packaging/common/cfengine-hub/postinstall.sh [hunk 3]
```diff
@@ -314,6 +314,7 @@ mkdir -p $CFENGINE_MP_DEFAULT_KEY_LOCATION
 mkdir -p $CFENGINE_MP_DEFAULT_CSR_LOCATION
 mkdir -p $CFENGINE_MP_DEFAULT_CERT_LINK_LOCATION
 CFENGINE_LOCALHOST=$(hostname -f | tr '[:upper:]' '[:lower:]')
+CFENGINE_LOCALIP=$(hostname -i)
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: On my machine:
```
$ hostname -i
fe80::99c7:97c9:edc7:7aff%enp3s0 fe80::7eef:e46f:949c:4a81%tun0 192.168.5.101 192.168.122.1 10.8.0.6
```
I guess we need to convert that into an `IP:...,IP:...` list?

**Review by vpodzime — COMMENTED** [review]
Please squash the commits into one or two with no intermediate introducing issues.

File: packaging/common/cfengine-hub/postinstall.sh (hunk 2 above)
Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh:343]: I don't want to include them twice and I always want to include `127.0.0.1` even if it is not mentioned from `hostname` commands.

File: packaging/common/cfengine-hub/postinstall.sh (hunk 3 above)
Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh]: I fixed this already. Thanks for checking.

**Comment by nickanderson** [comment]
@craigcomstock did you know there is a similar command in masterfiles? https://github.com/cfengine/masterfiles/blob/281e679230d9ac4273f6ec5e8774dc799f7c2809/cfe_internal/enterprise/mission_portal.cf#L375

File: packaging/common/cfengine-hub/postinstall.sh [hunk 4]
```diff
@@ -333,8 +336,14 @@ if [ ! -f $CFENGINE_MP_CERT ]; then
   # Generate a CSR in ${CFENGINE_MP_CSR} with key ${CFENGINE_MP_KEY}
   ${CFENGINE_OPENSSL} req -utf8 -sha256 -nodes -new -subj "/CN=$CFENGINE_LOCALHOST" -key ${CFENGINE_MP_KEY} -out ${CFENGINE_MP_CSR} ${OPENSSL_CNF}
 
+  # Build configuration with reasonable default subjectAltName entries
+  rm -f "$CFENGINE_MP_SSLCONF"
+  test -f "$OPENSSL_CNF" && cat "$OPENSSL_CNF" > $CFENGINE_MP_SSLCONF
+  CFENGINE_LOCALIP=$(for i in $(hostname -i; hostname -I); do echo $i; done | sed '/::1/d' | sed '/127.0.0.1/d' | sort -u | xargs -n1 printf "IP:%s\n" | paste -sd "," -)
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh:342]: You could combined the two `sed` runs into a single one using `-e ... -e ...`.

**Review by vpodzime — APPROVED** [review]
Looks good to me either way.

File: packaging/common/cfengine-hub/postinstall.sh (hunk 4 above)
Comment (olehermanse) [reply packaging/common/cfengine-hub/postinstall.sh:342]: TIL :) I like the simplicity of the pipes, do we have a performance benchmark? :)

Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh:342]: yeah, good tip. I do tend to like pipes for sake of super easy readability and composability, ideally I'd put all that on separate lines but that doesn't look too pretty. ;)

**Comment by craigcomstock** [comment]
cherry picks
https://github.com/cfengine/buildscripts/pull/1458
https://github.com/cfengine/buildscripts/pull/1457

**Pull request merged**